### PR TITLE
fix!: remove catch-all DSG redirect and handle discretely

### LINKED ISSUE #330: 404 pages don't work when a site has Functions but no SSR/DSG pages
Deploying a Gatsby 4 site with functions defined in the `api` folder breaks 404 handling in production.

If trying to access a non-existent page like `/404-test` triggers the below error: 
```
{"errorType":"Runtime.ImportModuleError","errorMessage":"Error: Cannot find module '/var/task/.cache/page-ssr'\nRequire stack:\n- /var/task/.netlify/functions-internal/__dsg/__dsg.js\n- /var/task/__dsg.js\n- /var/runtime/UserFunction.js\n- /var/runtime/index.js","trace":["Runtime.ImportModuleError: Error: Cannot find module '/var/task/.cache/page-ssr'","Require stack:","- /var/task/.netlify/functions-internal/__dsg/__dsg.js","- /var/task/__dsg.js","- /var/runtime/UserFunction.js","- /var/runtime/index.js","    at _loadUserApp (/var/runtime/UserFunction.js:202:13)","    at Object.module.exports.load (/var/runtime/UserFunction.js:242:17)","    at Object.<anonymous> (/var/runtime/index.js:43:30)","    at Module._compile (internal/modules/cjs/loader.js:1085:14)","    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)","    at Module.load (internal/modules/cjs/loader.js:950:32)","    at Function.Module._load (internal/modules/cjs/loader.js:790:12)","    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:76:12)","    at internal/main/run_main_module.js:17:47"]}
```

If you rename the gatsby `api` folder to `_api` and redeploy 404 pages start working correctly.

gatsby-plugin-netlify version 4.1 is installed and enabled.

### PR DESCRIPTION
**Pull request by orinokai**
### Summary

BREAKING CHANGE: DSG pages are now discretely redirected to the handler function, rather than via a catch-all redirect. This PR removes the catch-all redirect and moves DSG handling to [netlify/gatsby-plugin-netlify](https://github.com/netlify/gatsby-plugin-netlify), making that plugin a requirement for DSG pages.

This change ensures that 404 pages aren't unnecessarily handled via SSR and ensures they work when a site has functions but no SSR/DSG bundle.

### Test plan

1. The demo in the repo isn't able to provide a case for 'Functions but no SSR/DSG', so a new Gatsby test site is needed with that configuration.
2. 404 page should render successfully without being redirected to the handler function.

### Relevant links (GitHub issues, Notion docs, etc.) or a picture of cute animal

Fixes #330 

### Standard checks:

- [ ] Check the Deploy Preview's Demo site for your PR's functionality
- [ ] Add docs when necessary

---

🧪 Once merged, make sure to update the version if needed and that it was published correctly.

### CODE REVIEW

**Review by ericapisani — APPROVED** [review]
LGTM 🚀

**Review by ascorbic — APPROVED** [review]
Looks good, but let's hold off merging this until https://github.com/netlify/gatsby-plugin-netlify/pull/102 lands

File: plugin/src/helpers/config.ts [hunk 1]
```diff
@@ -65,14 +65,25 @@ function hasPlugin(plugins: PluginRef[], pluginName: string): boolean {
...
   const gatsbyConfig = loadGatsbyConfig({
     utils,
-    publish: netlifyConfig.build.publish,
+    gatsbyRoot,
   })
 
-  if (!hasPlugin(gatsbyConfig.plugins, 'gatsby-plugin-netlify')) {
+  if (hasPlugin(gatsbyConfig.plugins, 'gatsby-plugin-netlify')) {
+    if (
+      // prettier-ignore
```
Comment (orinokai) [inline plugin/src/helpers/config.ts:79]: FYI this ignore is just to avoid tripping over the max-lines lint rule!

Comment (ascorbic) [reply plugin/src/helpers/config.ts:79]: Why does it need a prettier ignore for that? Isn't that an eslint rule?

Comment (orinokai) [reply plugin/src/helpers/config.ts:79]: the file was under 150 lines until prettier formatted that 1 line onto 3. I tried ignoring prettier to see if it passed the eslint rules and it did, so I just kept it there. I can swap it for the eslint-ignore if you think it makes more sense?

Comment (ascorbic) [reply plugin/src/helpers/config.ts:79]: Yeah, just do the eslint ignore for max lines for the whole file

File: plugin/src/helpers/files.ts [hunk 2]
```diff
@@ -70,6 +78,26 @@ export const findModuleFromBase = ({ paths, candidates }): string | null => {
   return null
 }
 
+export async function checkPackageVersion(
+  root: string,
+  name: string,
+  version: string,
+): Promise<boolean> {
+  const packagePath = findModuleFromBase({
```
[the remark this answers is not in the document]
Comment (orinokai) [reply plugin/src/helpers/files.ts]: fair point, have amended

**Review by ascorbic — APPROVED** [review]
:shipit:

**Pull request merged**